charon/src/ast/types_utils.rs

```diff
@@ -819,6 +894,112 @@ impl Variant {
     }
 }
 
+impl DynPredicate {
+    /// Get the vtable declaration reference with current generics applied.
+    /// Matches associated types from the vtable's generics with the dyn predicates's constraints.
+    ///
+    /// Rustc guarantees all associated types are specified in a `dyn Trait` type.
+    pub fn vtable_ref(&self, translated: &TranslatedCrate) -> TypeDeclRef {
+        // Get vtable_ref's ID with trait-ref's generics from dyn-self applied.
+        // Add associated types in correct order following the vtable's generics.
+
+        // 0. Prepare trait name for debug/error messages
+        let trait_name = self.binder.params.trait_clauses[0]
+            .trait_
+            .skip_binder
+            .id
+            .with_ctx(&translated.into_fmt())
+            .to_string();
+
+        // 1. Get vtable ref from trait declaration
+        //    Provides: 1) final return ID, 2) correct order of associated types
+        // Firstly, get the trait declaration for the vtable ref it stores.
+        let trait_decl = translated
+            .trait_decls
+            .get(self.binder.params.trait_clauses[0].trait_.skip_binder.id)
+            .unwrap();
+
+        // Get vtable ref from definition for correct ID.
+        // Generics in vtable ref are placeholders but provide correct order of the associated types.
+        let Some(vtable_ref) = &trait_decl.vtable else {
+            panic!(
+                "Vtable for trait {} is None, meaning the trait is non-dyn-compatible!",
+                trait_name
+            );
```

Comment:
This panic is ok however, because if it happens we (charon developers) made a mistake somewhere.